lib/api/core/pluginsManager.js

```diff
@@ -0,0 +1,326 @@
+      }
+
+      if (installPlugins.call(this, customPlugins, defaultPlugins)) {
+        fs.writeFileSync(pathCustomPlugins, JSON.stringify(customPlugins, null, 2));
+      }
+    }
+
+    this.plugins = constructList(defaultPlugins, customPlugins);
+  };
+
+  /**
+   * Attach events hooks and pipes given by plugins
+   */
+  this.run = function () {
+    _.forEach(this.plugins, function (plugin) {
+
+      if (!plugin.activated) {
+        return true;
+      }
+
+      plugin.object.init(plugin.config, this.isDummy);
+
+      if (plugin.object.hooks) {
+        _.forEach(plugin.object.hooks, function (fn, event) {
+          if (plugin.object[fn]) {
+            kuzzle.on(event, function (message) {
+              plugin.object[fn](message, event);
+            });
+          }
+        });
+      }
+
+      if (plugin.object.pipes) {
+        _.forEach(plugin.object.pipes, function (fn, pipe) {
+          if (plugin.object[fn]) {
+            if (!this.pipes[pipe]) {
+              this.pipes[pipe] = [];
+            }
+
+            this.pipes[pipe].push(plugin.object[fn]);
+          }
+        }.bind(this));
+      }
+
+    }.bind(this));
+  };
+
+  /**
+   * Trigger an event for emit event and chain pipes
+   *
+   * @param event
+   * @param data
+   * @returns {Promise}
+   */
+  this.trigger = function (event, data) {
+    if (this.isDummy) {
+      this.kuzzle.emit(event, data);
+      return Promise.resolve(data);
+    }
+
+    return triggerPipes.call(this, event, data)
+      .then(function (modifiedData) {
+        return triggerHooks.call(this, event, modifiedData);
+      }.bind(this));
+  };
+};
+
+/**
+ * Emit event
+ * @param event
+ * @param data
+ * @returns {Promise}
+ */
+var triggerHooks = function (event, data) {
+  this.kuzzle.emit(event, data);
+
+  return Promise.resolve(data);
+};
+
+/**
+ * Chain call all attached functions plugins on the specific event
+ * @param event
+ * @param data
+ * @returns {Promise}
+ */
+var triggerPipes = function (event, data) {
+  var
+    deferred = q.defer(),
+    preparedPipes = [],
+    indexDelimiter,
+    wildcard;
+
+  indexDelimiter = event.indexOf(':');
+  if (indexDelimiter !== 1) {
+    wildcard = event.substring(0, indexDelimiter+1) + '*';
+  }
+
+  if (this.pipes && this.pipes[event] && this.pipes[event].length) {
+    preparedPipes = this.pipes[event];
+  }
+
+  if (wildcard && this.pipes && this.pipes[wildcard] && this.pipes[wildcard].length) {
+    preparedPipes = preparedPipes.concat(this.pipes[wildcard]);
+  }
+
+  if (preparedPipes.length > 0) {
+    // Add a fake function in first position for return the 'data' to other chained functions
+    preparedPipes.unshift(function (callback) { callback(null, data); });
+  }
+  else {
+    deferred.resolve(data);
+    return deferred.promise;
+  }
+
+  async.waterfall(preparedPipes, function (error, result) {
+    deferred.resolve(result);
```

Comment:
Proposal: to avoid badly coded plugins to freeze Kuzzle, maybe we should add a timeout and automatically reject the promise with an explicit error when it is reached?